Flows and shows Transitive Origin Data into PM UI

dominoFire opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/11486

**Regression?** No. 
**Last working version:** N/A

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->

Makes PM UI shows Transitive Packages data in PMUI, leveraging previously merged PR . Feature is behind experimentation flag. See animations below.

### Installed packages
![Installed packages group demo animation](https://user-images.githubusercontent.com/1192347/154588895-1e5942d6-dcaa-49ca-9b1b-e68459af4af1.gif)



### Transitive packages
![Transitive packages group demo animation](https://user-images.githubusercontent.com/1192347/154588911-7ff157bf-4ce8-4f69-8098-fab78a304dba.gif)


### Detailed changes list

- Adds a new implementation of `IPackageSearchMetadata` interface named `TransitivePackageSearchMetadata`, in order to flow transitive origins data to UI.
- Removes hardcoded string and unused logger in `InstalledPackageFeed` class.
- Adds new `IPackageFeed` implementation named `InstalledAndTransitivePackageFeed`, which handles installed and Transitive Packages with transitive origins data.
- Adds tests from new previously described feed.
- Flows new data into PM UI and sorts package list first by package level and then by package Id, ascending.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added: 1 new unit test class added with 4 tests
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - ~[ ] Test exception~
  - **OR**
  - ~[ ] N/A~ <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - ~[ ] Documentation PR or issue filled~
  - **OR**
  - [x] N/A: Changes are behind experimentation flag. Docs will filled when feature is rolled out.


## Accessibility Insights FastPass

![image](https://user-images.githubusercontent.com/1192347/154588799-82fbc814-f248-4ae7-95f6-e76a87f027c7.png)

Commit:
First working draft

- Wired to PM UI
- Sorting View Models in PM UI
- Updated comments
- Unseal and derive InstalledPackageFeed

Commit:
Tests

- Started tests
- Tests: more tests

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -292,15 +292,20 @@ public IEnumerable<PackageItemViewModel> GetCurrent()
+                {
+                    listItem.PackageLevel = metadata.TransitiveOrigins.Any() ? PackageLevel.Transitive : PackageLevel.TopLevel;
+                }
+
```
```suggestion
                    IncludePrerelease = _includePrerelease,
                    PackageLevel = metadata.TransitiveOrigins?.Any() == true ? PackageLevel.Transitive : PackageLevel.TopLevel
                };
```

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+
+            _installedPackages = installedPackages;
+            _transitivePackages = transitivePackages;
+            _metadataProvider = metadataProvider;
```
More concise:

```suggestion
            _installedPackages = installedPackages ?? throw new ArgumentNullException(nameof(installedPackages));
            _transitivePackages = transitivePackages ?? throw new ArgumentNullException(nameof(transitivePackages));
            _metadataProvider = metadataProvider ?? throw new ArgumentNullException(nameof(metadataProvider));
```

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+            _metadataProvider = metadataProvider;
+        }
+
+
```
Extra blank line

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+                throw new InvalidOperationException(Strings.Exception_InvalidContinuationToken);
+            }
+
+            // Remove foreign transitive packages
```
What are foreign transitive packages? I think those are the transitive dependencies through a project reference, is that correct? Could we explain it in this comment so it's easier to remember in the future if we look back at this code?

dominoFire replied:
Will add a comment summarizing this comment: https://github.com/NuGet/NuGet.Client/pull/4465#discussion_r809614157

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+                packageMetadata = await _metadataProvider.GetPackageMetadataAsync(identity, includePrerelease, cancellationToken);
+            }
+
+            IEnumerable<IPackageReferenceContextInfo> transPRs = identity.PackageReferences.Where(t => t is ITransitivePackageReferenceContextInfo);
```
Recommend a more descriptive variable name

```suggestion
            IEnumerable<IPackageReferenceContextInfo> transitivePRs = identity.PackageReferences.Where(t => t is ITransitivePackageReferenceContextInfo);
```

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+
+            IEnumerable<IPackageReferenceContextInfo> transPRs = identity.PackageReferences.Where(t => t is ITransitivePackageReferenceContextInfo);
+
+            IPackageReferenceContextInfo transPR = transPRs.OrderByDescending(x => x.Identity.Version).FirstOrDefault();
```
Same here, recommend more descriptive variable name
```suggestion
            IPackageReferenceContextInfo transitivePR = transitivePRs.OrderByDescending(x => x.Identity.Version).FirstOrDefault();
```

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+
+            IPackageReferenceContextInfo transPR = transPRs.OrderByDescending(x => x.Identity.Version).FirstOrDefault();
+
+            var transPRObject = transPR as TransitivePackageReferenceContextInfo;
```
```suggestion
            var transitivePRObject = transitivePR as TransitivePackageReferenceContextInfo;
```

jebriede replied:
I don't think we need this cast if we handle the list as per the comment above

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+            if (packageMetadata == null)
+            {
+                // and failing that we go to the network
+                packageMetadata = await _metadataProvider.GetPackageMetadataAsync(identity, includePrerelease, cancellationToken);
```
Is there a way to get metadata without pulling in deprecation and vulnerability data? For this iteration of the feature, we want to exclude the network calls to pull in deprecation and vulnerability data.

dominoFire replied:
Ok, now pulling metadata to only local sources. See edge case.

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+            IReadOnlyCollection<PackageIdentity> res = Array.Empty<PackageIdentity>();
+            if (transPRObject != null)
+            {
+                res = transPRObject.TransitiveOrigins?.Select(to => to.Identity)?.ToArray() ?? Array.Empty<PackageIdentity>();
```
If we ever change what res is initialized to, we'd have to change it in two places. Since res is already assigned to Array.Empty so we don't need to do that again.

dominoFire replied:
Refactored to initialize only once.

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+            if (res.Any())
+            {
+                var ts = new TransitivePackageSearchMetadata(packageMetadata, res);
+
```
extra line not needed

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+                packageMetadata = await _metadataProvider.GetPackageMetadataAsync(identity, includePrerelease, cancellationToken);
+            }
+
+            IEnumerable<IPackageReferenceContextInfo> transPRs = identity.PackageReferences.Where(t => t is ITransitivePackageReferenceContextInfo);
```
If we treat the list as a list of ITransitivePackageReferenceContextInfo's then we can avoid a lot of casting and it would help make the code more concise. We can also benefit from Linq's OfType here:

```suggestion
            IEnumerable<ITransitivePackageReferenceContextInfo> transitivePRs = identity.PackageReferences.OfType<ITransitivePackageReferenceContextInfo>().ToArray();
```

dominoFire replied:
Refactored.

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/ExperimentUtility.cs:
```diff
@@ -0,0 +1,28 @@
+            bool isExpEnabled;
+            try
+            {
+                var svc = await ServiceLocator.GetComponentModelServiceAsync<INuGetExperimentationService>();
```
It would be great to not have to locate the service for each call. The bool check should be cheap but with locating the service each time this call becomes more expensive.

dominoFire replied:
That's why it's inside an AsyncLazy. It is only called once, right?

jebriede replied:
Yep... I overlooked that by accident, my bad. 😆

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+            }
+
+            // Remove foreign transitive packages
+            IEnumerable<PackageCollectionItem> pkgsWithOrigins = _transitivePackages
```
What are foreign transitive packages?

dominoFire replied:
Foreign Transitive Package: A transitive package that is not brought by any installed package. Example

ProjectA has:
- PackageA that depends in T1, T2

ProjectB only has a project reference to ProjectA. Then:

ProjectA transitive packages with transitive origins:
- T1 (PackageA)
- T2 (PackageA) 

ProjectB transitive packages with transitive origins:
- T1 (none)
- T2 (none)
- ProjectA (none)

In this case, ProjectB has foreign transitive dependencies.

nkolev92 replied:
I agree with @jebriede. 

Please don't introduce a new term that doesn't have an obvious meaning. 

Just say "transitive packages from project references".

nkolev92 replied:
Are we excluding those packages from the view?

dominoFire replied:
Yes, "transitive packages from project references" will not be shown in PM UI Install Tab.

nkolev92 replied:
I don't think that's right. 

The design doesn't seem to call that out, https://github.com/NuGet/Home/blob/dev/proposed/2021/TransitiveDependenciesInPMUI.md. 

cc @chgill-MSFT @jebriede 

Why are we excluding packages that are brought in through projects?

jebriede replied:
The decision to show transitive packages for the current project and not for dependent projects of the current project was made with @chgill-MSFT because the scope is to show transitive dependencies on a project level, not solution level, related to this QA about viewing transitive packages in the project vs solution, here: https://github.com/NuGet/Home/blob/dev/proposed/2020/Transitive-Dependencies.md#qa

The functional spec should probably be updated to explicitly call out that transitive packages will be shown for the project being viewed in the PMUI and not cross project dependencies. To view transitive dependencies of another project, you'd view the installed packages in the PMUI for that project.

nkolev92 replied:
That doesn't seem like a great idea. :( 

Transitive packages brought in through other projects *are* the current project's transitive dependencies. 
You can upgrade them, tweak etc. 
It also doesn't really match dotnet list package.

dominoFire replied:
I think we should not include transitive dependencies brought in through projects because:

- What is the transitive origin of those packages? Should it be the project or the top-level package package? And, that top-level package is not in the project.
- If you chose to install that transitive dependency, would it make more sense to install it in the project with the top-level package?

nkolev92 replied:
I'd love to hear how other people on @NuGet/nuget-client feel about this.

nkolev92 replied:
> What is the transitive origin of those packages? Should it be the project or the top-level package package? And, that top-level package is not in the project.

It should be the project. If there's no top level package, there's no top level package. Who cares. Project/Package, same thing. 

> If you chose to install that transitive dependency, would it make more sense to install it in the project with the top-level package?

You can add a guidance about that easily.

nkolev92 replied:
Also sometimes, you may want a certain project to upgrade...ex. Executable vs Library.

jeffkl replied:
Team Review: Everyone should weigh in on whether or not transitive package references from project references should be included

jeffkl replied:
In my opinion, "transitive packages" in this context is every package that a particular project is consuming.  This should include packages flowing from projects.  Consider the scenario where my project has no package references but has one or more project references which pulls in packages.  This UI would show that my project has no top-level packages and no transitive packages right?  However, building my project would obviously show that I'm consuming packages.

chrisraygill replied:
Thanks for all the perspectives so far everyone!

I chose to only show transitive packages coming from top-level packages rather than projects for simplicity - to make sure actions like "promoting/installing" a transitive package and scope are clear.

My thought process is that packages from referenced projects should be handled in the PMUIs for those respective projects. However, I am open to other perspectives and potentially changing that design decision if it makes sense.

**Question 1:** Let's say we do surface transitive packages from referenced projects, so Project B --> Project A ---> Package A.

I'm using in the PMUI for Project B and see that transitive Package A has a vulnerability.

I'm assume that directly installing a non-vulnerable version of Package A in Project A is the ideal fix rather than directly installing it in Project B. Is that the case or am I missing anything?

**Question 2:** What are examples of problems/confusions we might introduce by not surfacing transitive packages from referenced projects?

FYI @JonDouglas

nkolev92 replied:
> I chose to only show transitive packages coming from top-level packages rather than projects for simplicity - to make sure actions like "promoting/installing" a transitive package and scope are clear.

Simplicity for promoting a package to top level a noble goal, but this is a "list" first feature, not an action first one. 
You can always put a little indicator that tells people that they may want to upgrade the package in the originating project instead and not the  current one. 

I'd argue what's not simple is not showing all the transitive packages to begin with. 

Think of the list here as "dotnet list package --include-transitive"

jeffkl replied:
@chgill-MSFT 

> I'm assume that directly installing a non-vulnerable version of Package A in Project A is the ideal fix rather than directly installing it in Project B. Is that the case or am I missing anything?

I think in that case, we should be smart enough to detect that the package actually came from Project A so we'd upgrade it in Project A.  At least if we show transitive packages, I'm more likely to see that I have a vulnerability somewhere in my dependency tree rather than having to go to each project to see what they have.

> Question 2: What are examples of problems/confusions we might introduce by not surfacing transitive packages from referenced projects?

One example is when Project B has no package references, only one or more project references.  I know that I'm using packages, but the UI shows me an empty list for "transitive packages".  But Solution Explorer shows me them, which I think is confusing.

Happy to chat about it so we're both on the same page!

dominoFire replied:
Assuming we can show all transitive dependencies, what would be the behavior on solution PM UI ?

JonDouglas replied:
@chgill-MSFT let's chat tomorrow and potentially meet with everyone to review shortly.

JonDouglas replied:
To echo @chgill-MSFT and also keep the project on-time with regards to delivery, here is another perspective.

1. The original design did take into account project references but was cut intentionally. The reason being is that the context of the user would be within the "project specific level PM UI" meaning the only context is the project packages and not it's project references. It would be confusing to have a transitive dependency be included from another project reference without a way to understand it's origin such as a UI affordance or hover experience. The goal with transitive packages is having a direct origin from their project's top-level package and not across project references to start. 
2. We should track a separate issue covering this design choice and with enough feedback post shipping this feature, we can iterate on including it with a new proposal specifically covering all flowing transitive packages across project boundaries. Including these packages in the list today would be confusing unless they had some special element inferring they came from another place outside the context of the project. @chgill-MSFT this would be a good issue to file.
3. Additionally, we may need to consider the current [.NET CLI experience](https://github.com/NuGet/Home/blob/dev/proposed/2020/Transitive-Dependencies.md#cli) to include origins which already support project references & the packages being brought in. Right now everything is just thrown into the list and it's not apparent where they came from. @chgill-MSFT this would also be a good issue to file for CLI <> IDE parity.

A lot of words to say these are all things we should consider while iterating further post release but shouldn't block on.

dominoFire replied:
Filled issue for point 2: https://github.com/NuGet/Home/issues/11624

Filled issue for point 3: https://github.com/NuGet/Home/issues/11625

nkolev92 replied:
Can someone please follow up and update the spec please? 
What is the behavior when a transitive package is coming through both package and project roots?

dominoFire replied:
> Can someone please follow up and update the spec please? 

Followed up with spec update PR: https://github.com/NuGet/Home/pull/11640

> What is the behavior when a transitive package is coming through both package and project roots?

It will show the transitive package with project-resolved version. Example:

Suppose you have a solution with two projects:

- ProjectA:
  * ProjectB
  * Newtonsoft.Json.Bson 1.0.2 (has dependency on Newtonsoft.Json 12.0.1) 
- ProjectB:
  * Microsoft.TestPlatform.TestHost 17.2.0-preview-20220131-20 (has dependency on Newtonsoft.Json 9.0.1)

PM UI on ProjectA

![image](https://user-images.githubusercontent.com/1192347/156535354-b8ec86e9-5ba5-4879-ae87-4171ff743c08.png)

PM UI on ProjectB

![image](https://user-images.githubusercontent.com/1192347/156535436-b7d56194-7591-46f8-88c9-53c97df9c647.png)

jebriede commented on src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/TransitivePackageSearchMetadata.cs:
```diff
@@ -0,0 +1,88 @@
+            }
+
+            _packageSearchMetadata = package;
+            TransitiveOrigins = transitiveOrigins;
```
```suggestion
            _packageSearchMetadata = package ?? throw new ArgumentNullException(nameof(package));
            TransitiveOrigins = transitiveOrigins ?? new ArgumentNullException(nameof(transitiveOrigins));
```

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,115 @@
+        private readonly IEnumerable<PackageCollectionItem> _transitivePackages;
+        private readonly IPackageMetadataProvider _metadataProvider;
+
+        public InstalledAndTransitivePackageFeed(IEnumerable<PackageCollectionItem> installedPackages, IEnumerable<PackageCollectionItem> transitivePackages, IPackageMetadataProvider metadataProvider)
```
You might be able to just inherit from the installedpackagefeed and reduce some duplication.

dominoFire replied:
InstalledPackageFeed is a sealed class. Can I un-seal it?

nkolev92 replied:
Probably?
UNless it's serialized, which pretty sure isn't, then it wouldn't make a difference. 

This is in NuGet.Clients, so it should be harmless.

dominoFire replied:
Refactored.

jebriede commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Feeds/InstalledAndTransitivePackageFeedTests.cs:
```diff
@@ -0,0 +1,171 @@
+                    return Task.FromResult(PackageSearchMetadataBuilder.FromIdentity(new PackageIdentity("packageA", NuGetVersion.Parse("1.0.0"))).Build());
+                });
+
+
```
extra blank line

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs:
```diff
@@ -373,7 +374,15 @@ private async ValueTask<IInstalledAndTransitivePackages> GetInstalledAndTransiti
-                packageFeeds.mainFeed = new InstalledPackageFeed(installedPackageCollection, metadataProvider, logger);
+                if (await ExperimentUtility.IsTransitiveOriginExpEnabled.GetValueAsync(cancellationToken))
+                {
+                    packageFeeds.mainFeed = new InstalledAndTransitivePackageFeed(installedPackageCollection, transitivePackageCollection, metadataProvider);
```
Can you calculate the transitivePackageCollection *only* when the experment is enabled?
Right now, this is an unnecessary perf regression.
Line 350.

dominoFire replied:
RecommenderFeed also needs `transitivePackageCollection`. That's why I left that code as-is.

nkolev92 replied:
oh my bad then. 

Resolving this comment as it doesn't apply.

jebriede reviewed (commented):
Great work! I like how this was handled with a specific feed that handles the logic of installed and transitive packages. Left some comments, a few styling suggestions, and a question about vulnerabilities and deprecations.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -292,15 +292,20 @@ public IEnumerable<PackageItemViewModel> GetCurrent()
 
+                if (metadata.TransitiveOrigins != null)
+                {
+                    listItem.PackageLevel = metadata.TransitiveOrigins.Any() ? PackageLevel.Transitive : PackageLevel.TopLevel;
```
Isn't TransitiveOrigins set for transitve packages only?

dominoFire replied:
Yes. That's why that list could be null for top-level, installed packages.

nkolev92 replied:
So no need to check for TransitiveOrigins.Any() then.

nkolev92 reviewed (commented):
A few questions.

Commit:
Tests:Feedback

Commit:
Defaulting to simple metadata when there is no available local metadata

Commit:
Feedback: null checks

Commit:
Surfaced packageFolders section from assets file to InstalledAndTransitivePackageFeed

Commit:
Fixed packageFolders NRE

erdembayar commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Common/IProjectContextInfoExtensions.cs:
```diff
@@ -68,6 +69,31 @@ public static class IProjectContextInfoExtensions
             }
         }
 
+        public static async ValueTask<IReadOnlyCollection<string>> GetPackageFoldersAsync(
```
Could you add at least 1 line of details on what kind of package folder it returns?
Someone looking at this code later need to spend time to understand what it's.

dominoFire replied:
Added docs.

Commit:
Docs and local metadata fixes

- Added docs for methods that surface packageFolders section from assets
  file
- Created method for retriving only local metadata without versions,
  deprecation and vulnerability info.

Commit:
Tests: Added more tests

chrisraygill commented:
Overall this looks great and I'm really happy with the progress!

A couple comments on the UI from the GIFs:

![image](https://user-images.githubusercontent.com/15097183/156016328-197a9c5e-d7e1-49ee-bee1-56e485770ce4.png)

1. Can we reduce the space between the "Top-level packages" group header and the top of the package list window? Looks a little awkward.
2. Will the count for the number of packages in each group come from a separate PR?

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Common/IProjectContextInfoExtensions.cs:
```diff
@@ -68,6 +69,40 @@ public static class IProjectContextInfoExtensions
+                throw new ArgumentNullException(nameof(serviceBroker));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using INuGetProjectManagerService projectManager = await GetProjectManagerAsync(serviceBroker, cancellationToken);
+
+            IReadOnlyCollection<string> packageFolders = await projectManager.GetPackageFoldersAsync(new string[] { projectContextInfo.ProjectId }, cancellationToken);
+
+            return packageFolders;
```
Can we reduce the number of blank lines here?

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Common/IProjectContextInfoExtensions.cs:
```diff
@@ -68,6 +69,40 @@ public static class IProjectContextInfoExtensions
+            return packageFolders;
+        }
+
+
```
Remove extra line

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -0,0 +1,84 @@
+                .Where(t => t.PackageReferences.Any(x => x is ITransitivePackageReferenceContextInfo y && y.TransitiveOrigins.Any()));
+
+            IEnumerable<PackageCollectionItem> pkgs = _installedPackages.Concat(pkgsWithOrigins);
+
```
Remove extra blank lines

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -316,6 +299,8 @@ private async ValueTask<IList<LockFileTarget>> GetTargetsListAsync(string assets
 
             LockFile lockFile = LockFileUtilities.GetLockFile(assetsFilePath, NullLogger.Instance);
 
```
Remove blank line above

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs:
```diff
@@ -326,6 +328,37 @@ private async ValueTask<IInstalledAndTransitivePackages> GetInstalledAndTransiti
+            IReadOnlyCollection<IProjectContextInfo> projectContextInfos,
+            CancellationToken cancellationToken)
+        {
+            Task<IReadOnlyCollection<string>>[] tasks = projectContextInfos.Select(pctxi => pctxi.GetPackageFoldersAsync(_serviceBroker, cancellationToken).AsTask()).ToArray();
+
+            IReadOnlyCollection<string>[] packageFolders = await Task.WhenAll(tasks);
+
+            HashSet<string> pkgFoldersUnique = new HashSet<string>();
+
+            packageFolders.ForEach(folders => pkgFoldersUnique.AddRange(folders));
+
+            IEnumerable<SourceRepository> assetsPackageFolders = pkgFoldersUnique.Select(folder => _sharedServiceState.SourceRepositoryProvider.CreateRepository(new PackageSource(folder)));
+
+            IEnumerable<SourceRepository> globalPackageFolderRepositories = await _globalPackageFolderRepositoriesLazy.GetValueAsync(cancellationToken);
+
```
That's a lot of blank lines. One blank line per line of code seems excessive. Can we group some lines of code and only put a blank line when we want to logically separate concepts in the code to make it easier to read?

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs:
```diff
@@ -578,6 +579,38 @@ public async ValueTask ExecuteActionsAsync(IReadOnlyList<ProjectAction> actions,
+                throw new ArgumentNullException(nameof(projectIds));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            IReadOnlyList<NuGetProject> projects = await GetProjectsAsync(projectIds, cancellationToken);
+
+            var prStyleTasks = new List<Task<IReadOnlyCollection<string>>>();
+
```
Blank lines

jebriede reviewed (approved):
Looks good, but please consider this approved with suggestions. The code could use some blank line cleanup. Then it looks good!

Commit:
Feedbac: Addressed newline comments

Commit:
Removed unused using

dominoFire commented:
> 1. Can we reduce the space between the "Top-level packages" group header and the top of the package list window? Looks a little awkward.

That space belongs to the yellow indicator that says 'Packages found' or 'Error happened'. Following up in another issue: https://github.com/NuGet/Home/issues/11641

> 2. Will the count for the number of packages in each group come from a separate PR?

Yes, see issue comment: https://github.com/NuGet/Home/issues/10750#issuecomment-1057851243

@chgill-MSFT
